### PR TITLE
[Model][PersonaPlex] Encode only consumed Mimi codebooks

### PR DESCRIPTION
**Pull request by LOGO127**
## Purpose

Implements **item 2** in #7389, independently of the other workstreams; this does not close the RFC.

`PersonaPlexMimiCodec.encode_frame()` currently runs Mimi's default 32-codebook quantizer and returns only the first 8. Pass `num_quantizers=CODEBOOKS` to avoid computing the discarded suffix. The returned shape/dtype/layout, checkpoint loading, decoder capacity, and streaming state remain unchanged. This does not touch the RoPE/ring-mask caching work in item 5.

The new checkpoint-free tests exercise the production `encode_frame()` with a small real streaming front end and the real HF Mimi quantizer. They verify exact prefix equality, that codebooks 9–32 are not executed, and parity across ring wrap, slot reset, and full stream reset at batch sizes 1, 2, and 9.

## Test Plan

**vLLM-Omni base:** `5a386d789f7646f42e7c97fe873848de9dd400aa`  
**Submitted commit:** `1dc64ad40eba26aa9bd015a0cd31bb646fd43bb9`  
**vLLM version:** `0.28.0` in the CPU test environment; not installed or used in the standalone GPU microbenchmark.

CPU test command in a compatible repository environment (no checkpoint required):

```bash
python -m pytest tests/model_executor/models/personaplex \
  -m 'core_model and cpu' --run-level=core_model -q
```

Local execution used Python 3.12.3, PyTorch 2.11.0+cpu, Transformers 5.14.1, and pytest 9.1.1. The existing local `cpu_pytest.py` wrapper sets `VLLM_TARGET_DEVICE=cpu` and makes only the NVML discovery helper unavailable; production modules, the quantizer, and pytest fixtures are imported normally. It is not included in this patch.

Standalone CUDA quantizer comparison, using the benchmark included in this PR:

```bash
python tests/model_executor/models/personaplex/benchmark_mimi_quantizer.py
```

For the installed ONNX/protobuf combination on the test machine, this was run with the process-local `PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python` environment variable. No packages or system configuration were changed.

## Test Result

- On the original base, the new module produces **3 failed / 3 passed**: the work-count guards fail because all 32 codebooks execute; the output-prefix checks do not indicate an existing output error.
- On the signed commit, the related PersonaPlex CPU directory produces **47 passed, 16 warnings**, including the 6 new cases.
- All applicable changed-file pre-commit gates pass, including Ruff, format, mypy for Python 3.10, SPDX, and test marks. `git diff --check` passes. The test is covered by the existing L1 model-executor test route.

**GPU microbenchmark:** one NVIDIA A100 80GB PCIe, driver 570.133.20, PyTorch `2.3.0a0+ebedce2`, CUDA runtime 12.3, Transformers 4.57.6, FP32. The default Mimi split-RVQ dimensions are used: hidden size 512, VQ dimension 256, 2048 centroids, and 32 codebooks. Centroids and latents are synthetic and seeded, not pretrained.

Each path has 16 explicit warmup calls, followed by 6 rounds of 16 calls with alternating path order. `torch.utils.benchmark.Timer` synchronizes accelerator work and adds its own warmup. Values below are per-call median [min, max] across those rounds, including host-launch time.

| Quantizer batch | Full 32-codebook encode (ms) | 8-codebook encode (ms) | Module speedup |
| --- | --- | --- | --- |
| 1 | 4.727 [4.677, 4.905] | 1.345 [1.327, 1.384] | 3.51x |
| 8 | 4.703 [4.678, 4.744] | 1.343 [1.327, 1.370] | 3.50x |
| 32 | 4.772 [4.747, 4.814] | 1.372 [1.355, 1.390] | 3.48x |

All **96 synthetic CUDA batch inputs** (32 per batch size) have bit-identical first-8 codebook indices. The benchmark emits all raw round timings and the HF quantizer source hash as JSON.

These are **quantizer-only** measurements, not full-codec/LM latency, audio-quality, GPU-memory reduction, or concurrent-session claims. No pretrained PersonaPlex checkpoint or end-to-end serving run was performed. CPU and GPU use the separate dependency versions listed above.

AI assistance: ChatGPT assisted with implementation, tests, benchmarks, and PR drafting.

### CODE REVIEW

**Comment by LOGO127** [comment]
Contributor-side self-review (AI-assisted), for unchanged commit `1dc64ad40`:

- Rechecked that `encode_frame` encodes only the eight consumed codebooks, with the same integer output, shape/layout, checkpoint loading and decoder capacity; no changes to RFC item 5.
- Reviewed the real-quantizer tests: exact first-eight prefix equality, no calls to codebooks 9–32, and parity across ring wrap, slot reset and full-stream reset.
- Rechecked the retained validation for this signed tree: **47 related CPU tests passed**, along with the applicable pre-commit gates. The earlier A100 run found bit-identical prefixes on **96 synthetic batch inputs**; the roughly **3.5x speedup is quantizer-only**, not whole-codec/LM or concurrent-serving performance.
- No pretrained PersonaPlex end-to-end/audio-quality validation is claimed. This follow-up rechecked source and retained evidence; it did not rerun CPU/GPU experiments or change the commit.

AI assistance: ChatGPT assisted with implementation, testing, benchmarks, and this self-review.

**Comment by LOGO127** [comment]
Additional correctness evidence for unchanged head `1dc64ad40`:

- With the bundled pretrained Mimi quantizer weights strictly loaded, 12 CPU FP32 cases (batches 1/2/8, four seeded synthetic latent inputs each) produced identical first-eight indices for full-32 versus eight-codebook encode.
- A separate real-audio codec check used the signed codec file and matching temporal dependency, with other imports from the existing environment (PyTorch 2.13.0+cu129, Transformers 5.14.1). Both paths received the same repository WAV, linearly resampled from 16 to 24 kHz. At batches 1 and 2, over 16 frame steps per path with a slot reset before frame 6 and a full reset before frame 11, encoded indices and decoded PCM were `torch.equal`.

The comparison reinitialized streaming state between paths and forced 32 quantizers only for the diagnostic baseline. These local diagnostic helpers are not included in this PR. They supplement the checkpoint-free tests; they do not establish long-stream/ring-wrap equivalence with pretrained weights, GPU equivalence, perceptual quality, or serving performance. The earlier synthetic GPU microbenchmark remains separate; no new speedup is claimed.

No code changes or additional scope in this update. AI assistance was used for these checks and this note.

**Review by linyueqian — APPROVED** [review]
Reviewed at head `1dc64ad40`. The production change is a provably equivalent one-liner, and the first test pins it in the right CI lane. Approving. Two nits below can land here or in a follow-up.

Why it is safe: in the transformers split quantizer, the single semantic codebook always runs in full and the acoustic RVQ iterates `layers[:num_quantizers - 1]` over successive residuals, so codes 0..7 cannot depend on layers 8..31. It is bit-identical by construction with no float reordering. `personaplex_mimi.py:395` is the only production caller, and the decode paths and Stage 0 consumer keep their shapes.

End to end: I served a stack of this PR with #7454, #7480, #7481, and #7448 on one H100 80GB against `main` at `e0c4f97f3` and ran the two-session lifecycle driver on both. The stack passed with statistics identical to the baseline (94/188/94 input frames, deficits 3/2/3, 19/38/19 audio chunks, voiced frames 22/57/22, RMS equal to ten decimals). Output audio was byte-identical as well: the three session WAVs hashed the same on the baseline, on a second baseline run, and on the stack. The PR's own test file passes in a vLLM 0.29.0 environment (6 passed).

1. `tests/model_executor/models/personaplex/benchmark_mimi_quantizer.py`: a 104-line CUDA-only microbenchmark in a unit-test directory, never collected (`python_files` is `test_*.py`). Repo convention for synthetic microbenchmarks is a subfolder under `benchmarks/` with a README (`benchmarks/README.md`, "Adding a new benchmark"). The numbers and the command already live in the PR body, so drop the script or move it.

2. `test_mimi_encode_codebooks.py:84-111`: the streaming-state assertions (conv carries, ring offsets, `_offset`) are redundant because the quantizer is stateless and runs after all of that state in `encode_frame`, so the codebook count cannot affect them. The output-code comparison across the slot reset and stream reset is fine to keep, but you can trim the state assertions.

Minor, no action: `_codec` duplicates the `__new__`-based checkpoint-free builder in `test_streaming_code2wav.py:131-151`, and `codes[:CODEBOOKS]` at `:396` is now a no-op slice.

Verified: marks `core_model, cpu` are collected by the ready-tier lane, and `pytest-mock` is pinned. There are no allowlist edits, no `torch.cuda.*` calls, and SPDX is present. The 3.5x quantizer-only number is a same-head toggle with warmup and a bit-identical guard, scoped clearly.

**Comment by Magi878** [comment]
Thanks for the approval and the two nits. I’ll keep the merge-critical scope unchanged: the production change is the one-line consumed-codebook limit with the focused correctness test. The benchmark-script placement and redundant streaming-state assertions can be cleaned up separately if preferred; neither changes the approved production behavior. Ready to merge from my side.

**Comment by linyueqian** [comment]
Holding the merge on CI for a moment. With the `ready` label, the AMD build (#11859) and NPU build (#6887) both failed on this head. The same two lanes passed on #7454 and #7481, which started in the same minute, so this is not lane-wide flakiness. Intel is red on every `main` commit right now and does not count.

The AMD ready lane is the only one that runs the new test (`pytest tests/model_executor -m 'core_model and cpu'`, sharded). That build stopped at 30 minutes against 66 for a passing run on the same lane, which looks like a shard dying early. The NPU ready lane runs no CPU tests, so a failure there should be unrelated to this diff, but I cannot read the job logs from my side.

@LOGO127 could you open the two builds and paste the failing job and its last assertion or traceback here? If it is the new test, check the transformers version on the ROCm image against the `MimiConfig` kwargs the test constructs (`vector_quantization_hidden_dimension`, `codebook_dim`, `num_semantic_quantizers`), then check whether `pytest-mock` is installed there. If both logs show something unrelated, a re-run is fine and I will merge once AMD is green.

**Comment by LOGO127** [comment]
@linyueqian I rechecked the unchanged head `1dc64ad40`: the main Buildkite status is successful, while AMD #11859 and NPU #6887 still report failure. I have not been able to retrieve either build's job logs from the available access, so I cannot yet identify the failing shard or classify the failures as unrelated. No speculative code change or rerun has been made.

This status check and follow-up were prepared with ChatGPT; no new test execution is claimed.

Could someone with access to the AMD/NPU pipelines attach the failing job name and last assertion/traceback here? For an AMD failure in `test_mimi_encode_codebooks.py`, the installed Transformers version and whether the `mocker` fixture is available would resolve the two checks you identified. I am keeping the approved production change and optional cleanup unchanged while that evidence is missing.

**Comment by linyueqian** [comment]
A cheaper next step than reading logs: this head sits on a Sept 10 base, and both failing lanes run tests this diff does not touch. NPU is green on the six newest `main` commits, and `main` merged a MiniCPM-o NPU code2wav stabilization fix this afternoon. #7480, which carries current `main`, passed NPU an hour ago. The AMD lane does run the new CPU test, so a rebase settles that question too.

@LOGO127 please rebase onto current `main` and push, and re-add the `ready` label if it drops. If AMD or NPU still fails on the fresh base, paste the failing job's last lines here. The approval stands.

**Comment by LOGO127** [comment]
Updated the branch onto current `main` at `538eddd74d57ac5a8fea0a8ac0d5288e20678ccb` without changing the reviewed PR diff. New head: `eada51617541d9be95ab2f3a88e66e4d55e9c3e3`.

I verified the branch is now 0 commits behind `main`, the diff is still the same three files, and the `ready` label is still present. Fresh CI has started; if AMD or NPU still fails on this base, I’ll post the failing job’s last lines rather than speculate.

**Comment by LOGO127** [comment]
Updated the branch onto current `main` at `f3f8ebfc25de` with new head `61a5645b85bd`. The branch is now 0 commits behind `main`, the `ready` label is still present, and the PR diff remains the same three files.

I resolved the upstream PersonaPlex codec changes by retaining the new active-mask streaming path and reapplying only the approved `num_quantizers=CODEBOOKS` optimization at the quantizer call. Fresh CI has started. If a vendor lane fails again, I’ll inspect the failing job before changing code.

AI assistance was used for this branch refresh and verification.
